### PR TITLE
Move dynamic mapping parameter configuration to @Document and @Field

### LINKED ISSUE #1871: Move dynamic mapping parameter configuration to @Document and @Field
At the moment you have to add a separate annotation `@DynamicMapping` to the entity or fields of type `Object` and `Nested` to configure the dynamic mapping parameter. This seems a bit too complicated to me, since the other mapping parameters like `ignore_above` or `doc_values` are part of the `@Field` annotation. We should add `dynamic` as part of the `@Document` and `@Field` annotation, deprecate the old implementation and switch to the new one.

### PR DESCRIPTION
**Pull request by xhaggi**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] **There is a ticket in the bug tracker for the project in our [issue tracker](https://github.
  com/spring-projects/spring-data-elasticsearch/issues)**. Add the issue number to the _Closes #issue-number_ line below
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Closes #1871

### CODE REVIEW

File: src/main/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentEntity.java [hunk 1]
```diff
@@ -102,6 +104,7 @@ public SimpleElasticsearchPersistentEntity(TypeInformation<T> typeInformation,
 			this.indexName = document.indexName();
 			this.versionType = document.versionType();
 			this.createIndexAndMapping = document.createIndex();
+			this.dynamic = document.dynamic();
 		}
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentEntity.java:109]:
```suggestion
			this.dynamic = document.dynamic();
		} else {
                       this.dynamic = Dynamic.INHERIT;
               }
```

see above

File: src/main/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentEntity.java [hunk 2]
```diff
@@ -73,6 +74,7 @@
 	private @Nullable ElasticsearchPersistentProperty joinFieldProperty;
 	private @Nullable VersionType versionType;
 	private boolean createIndexAndMapping;
+	private Dynamic dynamic = Dynamic.INHERIT;
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentEntity.java:77]:
```suggestion
	private final Dynamic dynamic = Dynamic;
```

I'd prefer a `final`property

**Review by sothawo — CHANGES_REQUESTED** [review]
some minor remarks, a failing test class

File: src/main/java/org/springframework/data/elasticsearch/core/index/MappingBuilder.java [hunk 3]
```diff
@@ -197,7 +197,12 @@ private void mapEntity(XContentBuilder builder, @Nullable ElasticsearchPersisten
 			}
 		}
 
-		if (dynamicMapping != null) {
+		if (isRootObject && entity != null && entity.dynamic() != Dynamic.INHERIT) {
+			builder.field(TYPE_DYNAMIC, entity.dynamic().name().toLowerCase());
+		} else if (nestedOrObjectField && parentFieldAnnotation != null
+				&& parentFieldAnnotation.dynamic() != Dynamic.INHERIT) {
+			builder.field(TYPE_DYNAMIC, parentFieldAnnotation.dynamic().name().toLowerCase());
```
[the remark this answers is not in the document]
Comment (xhaggi) [reply src/main/java/org/springframework/data/elasticsearch/core/index/MappingBuilder.java:204]: The method is called twice, for the root and for a nested entity and I thought it was necessary, but it's not. To make both work, I simply removed the "isRootObject" from the "if" statement.

**Review by sothawo — APPROVED** [review]

**Pull request merged**